bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/types/BUnionType.java

```diff
@@ -114,49 +93,34 @@ public BUnionType(List<Type> memberTypes, boolean readonly) {
         this(memberTypes, 0, readonly, false);
     }
 
-    public BUnionType(List<Type> memberTypes, boolean readonly, boolean isCyclic, Set<Type> unresolvedTypes) {
-        this(memberTypes, 0, readonly, isCyclic, unresolvedTypes);
-    }
-
-    public BUnionType(String typeName, Module pkg, boolean readonly, Class<? extends Object> valueClass) {
-        super(typeName, pkg, valueClass);
+    public BUnionType(List<Type> memberTypes, boolean readonly, boolean isCyclic) {
+        super(null, null, Object.class);
+        this.typeFlags = 0;
         this.readonly = readonly;
+        setMemberTypes(memberTypes);
+        this.isCyclic = isCyclic;
     }
 
-    public BUnionType(Type[] memberTypes, int typeFlags, boolean readonly, boolean isCyclic) {
-        this(Arrays.asList(memberTypes), typeFlags, readonly, isCyclic);
+    public BUnionType(Type[] memberTypes, Type[] originalMemberTypes, int typeFlags) {
+        this(memberTypes, originalMemberTypes, typeFlags, false, 0);
     }
 
-    public BUnionType(Type[] memberTypes, int typeFlags) {
-        this(memberTypes, typeFlags, false, false);
+    public BUnionType(Type[] memberTypes, Type[] originalMemberTypes, int typeFlags, boolean isCyclic, long flags) {
+        this(Arrays.asList(memberTypes), Arrays.asList(originalMemberTypes), typeFlags, isCyclic, flags);
     }
 
-    /**
-     * Constructor used when defining union type defs where cyclic reference is possible.
-     *
-     * @param typeFlags flags associated with the type
-     * @param readonly boolean represents if the type is readonly
-     * @param isCyclic boolean represents if the type is cyclic
-     */
-    public BUnionType(int typeFlags, boolean readonly, boolean isCyclic) {
-        super(null, null, Object.class);
+    public BUnionType(String name, Module pkg, int typeFlags, boolean isCyclic, long flags) {
+        super(name, pkg, Object.class);
         this.typeFlags = typeFlags;
-        this.readonly = readonly;
-        this.memberTypes = new ArrayList<>(2);
+        this.readonly = isReadOnlyFlagOn(flags);
+        this.memberTypes = null;
```

Comment:
Can't this lead to NPEs? Shouldn't we have an empty list instead?

Reply:
Yeah, updated in a422002, thanks!